### PR TITLE
Fix #6050: Crash on placing scenery clusters on water near map edge

### LINKED ISSUE #6050: Crash when placing scenery clusters near water's edge in the scenario editor
**OS:** Windows 7
**Version:** 0.1.1
**Commit/Build:** c595d24

**Steps to reproduce:**
1. Open the scenario editor
2. Proceed to the landscape editor
3. Generate a map with a large body of water **edit: on the edge of the map**
4. Using the cluster tool, select some scenery and click in the water a few times
5. _**Crash.**_

**Dump file:**
[8fecb482-c595-4b34-b0d5-9d43d2a824ba(c595d24_x86-64).zip](https://github.com/OpenRCT2/OpenRCT2/files/1181523/8fecb482-c595-4b34-b0d5-9d43d2a824ba.c595d24_x86-64.zip)

**Error log:**
`ERROR[c:\users\noah\documents\github\openrct2\src\openrct2\world\map.c:219 (map_get_first_element_at)]: Trying to access element outside of range`

**Exception details:**
`Exception thrown: read access violation.
mapElement was nullptr. occurred`

Screenshot to help with reproduction Screenshot to help with reproduction

### PR DESCRIPTION
**Pull request by rwjuk**
Caused by dereferencing null pointer in `map_get_water_height()`.

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
I don't think returning zero is the best idea it may make other issues.

**Comment by Gymnasiast** [comment]
Normally we'd return MONEY32_UNDEFINED in such circumstances.

**Comment by IntelOrca** [comment]
No please, don't do the null check here - map element not null is a pre-condition of a method like this. The callers should be the ones to check.

For example if we change this later to accept a `const rct_map_element &` you would not be able to check for null because its a (should be) non-nullable reference.

**Comment by IntelOrca** [comment]
There is another case of `map_get_water_height(surfaceElement)` further below in the function.

File: src/openrct2/world/SmallScenery.cpp [hunk 1]
```diff
@@ -322,7 +322,7 @@ static money32 SmallSceneryPlace(sint16 x,
 
         if (!isOnWater)
         {
-            if (map_get_water_height(surfaceElement) ||
+            if (surfaceElement != NULL && map_get_water_height(surfaceElement) ||
                 (surfaceElement->base_height * 8) != targetHeight)
```
Comment (IntelOrca) [inline src/openrct2/world/SmallScenery.cpp]: You need another check on the RHS of this OR clause.

**Comment by IntelOrca** [comment]
Line 311: `(surfaceElement->properties.surface.slope & 0x1F))`

Another check for this too.

File: src/openrct2/world/SmallScenery.cpp (hunk 1 above)
Comment (rwjuk) [inline src/openrct2/world/SmallScenery.cpp]: I moved the NULL check to another `if` above - no need to do the check twice.

**Comment by rwjuk** [comment]
Hang on, I'll fix whatever went wrong there...

**Pull request merged**